vendor: Add --skip-path-deps to skip path dependencies

stevenroose opened the pull request:
This changes the erroneous behavior to skip path dependencies and adds a
command option to keep the current behavior.

Fixes https://github.com/rust-lang/cargo/issues/10134 and https://github.com/rust-lang/cargo/issues/9172.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @ehuss (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

stevenroose commented:
I can't build right now, not sure if my fault or not..

```
$ cargo build
   Compiling sized-chunks v0.6.5
   Compiling serde_json v1.0.39
   Compiling tempfile v3.2.0
   Compiling semver v1.0.4
   Compiling toml v0.5.8
   Compiling os_info v3.0.8
   Compiling serde_ignored v0.1.2
   Compiling cargo-platform v0.1.2 (/home/steven/code/rust/cargo/crates/cargo-platform)
   Compiling crypto-hash v0.3.3
   Compiling cargo v0.59.0 (/home/steven/code/rust/cargo)
   Compiling im-rc v15.0.0
   Compiling git2 v0.13.24
   Compiling cargo-util v0.1.1 (/home/steven/code/rust/cargo/crates/cargo-util)
error: failed to run custom build command for `cargo v0.59.0 (/home/steven/code/rust/cargo)`

Caused by:
  process didn't exit successfully: `/media/data/cargo-target/debug/build/cargo-c079119e759e589a/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at 'attempted to zero-initialize type `ffi::imp::z::z_stream`, which is invalid', /home/steven/.cargo/registry/src/github.com-1ecc6299db9ec823/flate2-1.0.6/src/ffi.rs:79:42
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panicking.rs:517:5
     1: core::panicking::panic_fmt
               at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/panicking.rs:101:14
     2: core::panicking::panic
               at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/panicking.rs:50:5
     3: core::mem::zeroed
               at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/mem/mod.rs:629:9
     4: <flate2::ffi::imp::StreamWrapper as core::default::Default>::default
               at /home/steven/.cargo/registry/src/github.com-1ecc6299db9ec823/flate2-1.0.6/src/ffi.rs:79:42
     5: flate2::mem::Compress::new
               at /home/steven/.cargo/registry/src/github.com-1ecc6299db9ec823/flate2-1.0.6/src/mem.rs:202:29
     6: flate2::gz::write::gz_encoder
               at /home/steven/.cargo/registry/src/github.com-1ecc6299db9ec823/flate2-1.0.6/src/gz/write.rs:48:36
     7: flate2::gz::GzBuilder::write
               at /home/steven/.cargo/registry/src/github.com-1ecc6299db9ec823/flate2-1.0.6/src/gz/mod.rs:175:9
     8: build_script_build::compress_man
               at ./build.rs:17:19
     9: build_script_build::main
               at ./build.rs:7:5
    10: core::ops::function::FnOnce::call_once
               at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

stevenroose commented:
Added test, fixed build and optimized something I oversaw.

I can understand that it might be good behavior to not automatically duplicate path dependencies that are inside the current path.

So let's look at some different options for behavior:

1. never vendor path deps _(current behavior pre-patch)_
2. never vendor path deps, but add option to vendor all path deps
3. always vendor all path deps, but add option to skip vendoring _(behavior in this patch right now)_
4. vendor all path deps outside of current dir, don't vendor any path deps inside current dir
5. by default, vendor all path deps outside of current dir and don't vendor any path deps inside current dir, but have flags to vendor all/none

stevenroose commented:
Currently, it still suffers from the `.cargo/config` output being filled with absolute paths instead of relative ones.

bors commented:
:umbrella: The latest upstream changes (presumably #10161) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented:
In its current form, this would break backwards compatibility, by changing the existing behavior. That's not something we can do lightly; it would take a transition of some kind, assuming we wanted to make that change.

Given that, based on our conversation in today's @rust-lang/cargo meeting, I'm going to close this PR. However, we'd be happy to talk through details and use cases more on #10134.

hbina commented:
Would you be opposed to adding an option to do the reverse? i.e. include local paths?

stevenroose commented:
I pushed an updated version of this branch that is rebased on master.
